src/bounded_queue.ml

```diff
@@ -0,0 +1,113 @@
+(*
+  structure node t value: data type, next: pointer to node t
+  structure queue t Head: pointer to node t, Tail: pointer to node t, H lock: lock type, T lock: lock type
+*)
+
+type 'a node = Nil | Next of 'a * 'a node ref
+
+type 'a t = {
+  mutable head : 'a node ref;
+  mutable tail : 'a node ref;
+  head_lock : Mutex.t;
+  tail_lock : Mutex.t;
+  not_empty : Condition.t;
+  not_full : Condition.t;
+  size : int Atomic.t;
+  capacity : int;
+}
+
+(* create and initialize the 2 lock queue *)
+let create () =
+  let dummy = ref Nil in
+  {
+    head = dummy;
+    tail = dummy;
+    head_lock = Mutex.create ();
+    tail_lock = Mutex.create ();
+    not_empty = Condition.create ();
+    not_full = Condition.create ();
+    size = Atomic.make 0;
+    capacity = 1_000_000;
+  }
+
+(* push element to tail of queue *)
+let push t value =
+  let new_tail = ref Nil in
+  let new_node = Next (value, new_tail) in
+  Mutex.lock t.tail_lock;
+  (* Ensures queue is not full *)
+  while Atomic.get t.size = t.capacity do
+    Condition.wait t.not_full t.tail_lock
+  done;
+  (* Check if queue is empty before this insertion *)
+  t.tail := new_node;
+  t.tail <- new_tail;
+  let empty = Atomic.fetch_and_add t.size 1 = 0 in
+  Mutex.unlock t.tail_lock;
+  (* If it was empty then signal the waiting pop threads to wake up *)
+  if empty then (
+    Mutex.lock t.head_lock;
+    Condition.broadcast t.not_empty;
+    Mutex.unlock t.head_lock)
```

Comment:
This can be optimized to:

```ocaml
    Mutex.lock t.head_lock;
    Mutex.unlock t.head_lock;
    Condition.broadcast t.not_empty (* or signal *)
```

Why does this work and why is the `Mutex.lock - unlock` even needed?

The key is that a condition variable has no memory, so we must ensure that any potential waiter has called `Condition.wait` before we call `Condition.broadcast` (or `Condition.signal`) or reads the updated (non-zero) `t.size`.

Take a look at the `wait` in `pop`:

```ocaml
  Mutex.lock t.head_lock;
  (* Ensures queue is not empty *)
  while Atomic.get t.size = 0 do
    Condition.wait t.not_empty t.head_lock
  done;
```

It is possible that, if the signaler doesn't take the mutex, then the signal could happen after `Atomic.get` and before `Condition.wait` (and the `wait` might never return).  However, if the signaler takes the lock, i.e. calls `Mutex.lock t.head_lock`, then it must be that either the `Mutex.lock` returns after the `Condition.wait` or it returns before the `Mutex.lock` of a potential `Condition.wait` call.  In either case things will work out.  In first case the `Condition.broadcast` will wake up the waiter.  In the second case the `Atomic.get` before wait will return a non-zero value.

Reply:
done